File: src/server/src/main/java/com/project/application/services/CategoryService.java
package com.project.application.services;

import com.project.application.models.category.CategoryViewResource;
import com.project.domain.category.Category;
import com.project.infrastructure.data.CategoryRepository;
import io.micronaut.transaction.annotation.Transactional;
import jakarta.inject.Singleton;
import jakarta.persistence.EntityNotFoundException;

import java.util.List;
import java.util.UUID;

@Singleton
public class CategoryService {
    private final CategoryRepository categoryRepository;

    public CategoryService(CategoryRepository categoryRepository) {
        this.categoryRepository = categoryRepository;
    }

    @Transactional(readOnly = true)
    public List<CategoryViewResource> getCategories() {
        return this.categoryRepository.findAll()
                .stream()
                .map(CategoryViewResource::from)
                .toList();
    }

    @Transactional(readOnly = true)
    public Category getCategoryBy(UUID id) {
        return this.categoryRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(String.format("Category with %s does not exist", id)));
    }
}

File: src/server/src/main/java/com/project/application/services/CategoryService.kt
package com.project.application.services

import com.project.application.models.category.CategoryViewResource
import com.project.domain.category.Category
import com.project.infrastructure.data.CategoryRepository
import io.micronaut.transaction.annotation.Transactional
import jakarta.inject.Singleton
import jakarta.persistence.EntityNotFoundException
import java.util.UUID

@Singleton
open class CategoryService(private val categoryRepository: CategoryRepository) {
    @Transactional(readOnly = true)
    open fun getCategories(): List<CategoryViewResource> {
        return categoryRepository.findAll()
                .map { category: Category -> CategoryViewResource.from(category) }
    }

    @Transactional(readOnly = true)
    open fun getCategoryBy(id: UUID): Category {
        return categoryRepository.findById(id)
                .orElseThrow { EntityNotFoundException(String.format("Category with %s does not exist", id)) }
    }
}
